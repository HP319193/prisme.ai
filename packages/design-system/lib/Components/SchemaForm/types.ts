import { FunctionComponent, ReactNode } from 'react';
import { DatePickerProps } from '../DatePicker';
import { TextAreaProps } from '../TextArea';
import { TagsOption } from '../Select';

export const schemaTypes = [
  'string',
  'localized:string',
  'number',
  'localized:number',
  'boolean',
  'localized:boolean',
  'object',
  'array',
] as const;
export type SchemaTypes = typeof schemaTypes[number];

export const defaultUploadAccept =
  'image/gif,image/jpeg,image/png,image/svg+xml,' as const;

export const UIWidgetsForString = [
  'upload',
  'textarea',
  'date',
  'select',
  'color',
  'autocomplete',
  'radio',
  'html',
  'slider',
  'code',
] as const;
export const UIWidgetsForLocalizedString = ['textarea'] as const;
export const UIWidgetsForNumber = ['slider'] as const;
export const UIWidgetsForArray = ['tags', 'upload'] as const;
export const UIWidgetsByType = {
  [schemaTypes[0]]: UIWidgetsForString,
  [schemaTypes[1]]: UIWidgetsForLocalizedString,
  [schemaTypes[2]]: UIWidgetsForNumber,
  [schemaTypes[3]]: UIWidgetsForNumber,
  [schemaTypes[7]]: UIWidgetsForArray,
};

type StringWithAutocomplete<T> = T | (string & {});
export type UIWidgets = StringWithAutocomplete<
  typeof UIWidgetsForString[number]
>;

export type UiOptionsCommon = {
  field: {
    updateValue?: false | 'blur';
  };
};

export type UIOptionsNumber = {
  number: {
    step?: number;
  };
};
export type UiOptionsGrid = {
  grid: string[][][];
};
export type UiOptionsOneOf = {
  oneOf: {
    options: {
      label: string;
      index: number;
      value?: Record<string, any>;
    }[];
  };
};
export type UiOptionsArray = {
  array: 'row';
};

export type UiOptionsTextArea = {
  textarea: Omit<TextAreaProps, 'value'>;
};
export type UiOptionsUpload = {
  upload: {
    // https://developer.mozilla.org/fr/docs/Web/HTML/Element/Input/file#accept
    accept?: string;
    defaultPreview?: string | ReactNode;
  };
};
export type UiOptionsTags = {
  tags: {
    allowNew?: boolean;
    options: TagsOption[];
  };
};
export type UiOptionsSelect = {
  select: {
    options?: {
      label: string | ReactNode;
      value: any;
    }[];
    hideSearch?: boolean;
  };
};
export type UiOptionsDynamicAutocomplete = {
  autocomplete: string;
};
export type UiOptionsAutocomplete = {
  autocomplete: {
    options: {
      label: string | ReactNode;
      value?: any;
      options?: { label: string | ReactNode; value?: any }[];
    }[];
    minChars?: number;
  };
};
export type UiOptionsDate = {
  date: DatePickerProps;
};
export type UiOptionsHTML = {
  html: {
    htmlModeOnly?: boolean;
  };
};
export type UiOptionsSlider = {
  slider: {
    steps: {
      label: string;
      description: string;
      value: string | number;
      className?: string;
    }[];
    step?: number;
    showTooltip?: 'always' | 'hover';
  };
};
export type UiOptionsCode = {
  code: {
    mode: 'json' | 'css' | 'html' | 'javascript' | 'yaml';
  };
};
export interface Schema extends Record<string, any> {
  // Field type
  type?: SchemaTypes;
  // Field label
  title?: string | ReactNode;
  // Description in a help tooltip
  description?: string | ReactNode;
  // Input placeholder
  placeholder?: string;
  // Add button label when in a list of items or properties
  add?: string;
  // Remove button label when in a list of items or properties
  remove?: string;
  // Property key label
  propertyKey?: string;
  // Property value label
  propertyValue?: string;
  // For object type, object properties
  properties?: Record<string, Schema>;
  // Schema of object additional properties
  additionalProperties?: boolean | Schema;
  // For array type, schema of array items
  items?: Schema;
  // List of different schemas available
  oneOf?: Schema[];
  // Value of the oneOf field.
  value?: any;
  // List of constrainted values
  enum?: any[];
  // Labels of constrainted values
  enumNames?: string[];
  // Default value
  default?: any;
  // Do not display the field
  hidden?: true;
  // Disable the field
  disabled?: true;
  // Validation by regexp
  // @deprecated
  pattern?: string;
  // Validation errors messages
  // @deprecated
  errors?: Record<string, string>;
  validators?: Record<
    'required' | 'pattern' | 'min' | 'max' | 'tel' | 'email' | 'date',
    | {
        value?: string | number | boolean;
        message?: string;
      }
    | true
  >;
  // Custom widget
  'ui:widget'?: UIWidgets | FunctionComponent<FieldProps>;
  // Options for UI components
  'ui:options'?:
    | UiOptionsCommon
    | UIOptionsNumber
    | UiOptionsGrid
    | UiOptionsOneOf
    | UiOptionsArray
    | UiOptionsTextArea
    | UiOptionsUpload
    | UiOptionsSelect
    | UiOptionsTags
    | UiOptionsDate
    | UiOptionsAutocomplete
    | UiOptionsDynamicAutocomplete
    | UiOptionsHTML
    | UiOptionsSlider
    | Record<string, any>;
}

export interface FieldProps {
  schema: Schema;
  name: string;
  label?: string;
  className?: string;
}
