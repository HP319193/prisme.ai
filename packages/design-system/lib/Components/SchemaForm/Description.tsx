import { InfoCircleOutlined } from '@ant-design/icons';
import { Tooltip } from 'antd';
import { FC } from 'react';

interface DescriptionProps {
  text?: string;
  className?: string;
}

export const Description: FC<DescriptionProps> = ({
  children,
  text,
  className,
}) => {
  if (!text) return <>{children}</>;

  return (
    <div className={`relative w-full ${className}`}>
      {children}
      <Tooltip title={text} placement="left">
        <button type="button" className="absolute top-0 right-2">
          <InfoCircleOutlined />
        </button>
      </Tooltip>
    </div>
  );
};

export default Description;
