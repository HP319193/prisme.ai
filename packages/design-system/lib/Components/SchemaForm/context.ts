import { UiOptionsSelect, Schema } from '@prisme.ai/design-system';
import { createContext, FC, ReactElement, useContext } from 'react';
import { SelectProps } from '../Select';
import { DefaultFieldContainer } from './FieldContainer';
import { FieldProps, UiOptionsAutocomplete, UiOptionsHTML } from './types';

export type FieldComponent<T = any> = (
  props: FieldProps & T
) => ReactElement | null;
export type InputComponent<T = {}> = (
  props: T & {
    value: string;
    onChange: (e: React.ChangeEvent<HTMLInputElement> | string) => void;
  }
) => ReactElement;

export interface SchemaFormContext {
  locales: {
    submit?: string;
    addProperty?: string;
    removeProperty?: string;
    propertyKey?: string;
    propertyValue?: string;
    addItem?: string;
    removeItem?: string;
    oneOfOption?: string;
    uploadLabel?: string;
    uploadRemove?: string;
    freeAdditionalPropertiesLabel?: string;
  };
  components: {
    FieldContainer: FC<FieldProps>;
    FieldLocalizedText?: FieldComponent;
    FieldText?: FieldComponent;
    FieldBoolean?: FieldComponent;
    FieldLocalizedBoolean?: FieldComponent;
    FieldObject?: FieldComponent;
    FieldArray?: FieldComponent;
    FieldAny?: FieldComponent;
    FieldFreeAdditionalProperties?: FieldComponent;
    FieldSelect?: FieldComponent<{ options?: SelectProps['selectOptions'] }>;
    FieldRadio?: FieldComponent<{ options?: SelectProps['selectOptions'] }>;
    FieldDate?: FieldComponent;
    JSONEditor?: InputComponent;
    HTMLEditor?: InputComponent<{ options?: UiOptionsHTML }>;
    FreeAdditionalProperties?: FieldComponent;
    ManagedAdditionalProperties?: FieldComponent;
    UiWidgets: Record<string, FieldComponent>;
  };
  utils: {
    extractSelectOptions: (
      schema: Schema
    ) => UiOptionsSelect['select']['options'] | null;
    extractAutocompleteOptions: (
      schema: Schema
    ) => UiOptionsAutocomplete['autocomplete']['options'] | null;
    uploadFile: (
      base64File: string
    ) => Promise<
      string | { value: string; preview: string | ReactElement; label?: string }
    >;
  };
}

export const FieldContainer = DefaultFieldContainer;
export const context = createContext<SchemaFormContext>({
  locales: {},
  components: {
    FieldContainer: DefaultFieldContainer,
  },
  utils: {
    extractSelectOptions: () => [],
    extractAutocompleteOptions: () => [],
    uploadFile: async () => '',
  },
});

export const useSchemaForm = () => useContext(context);
