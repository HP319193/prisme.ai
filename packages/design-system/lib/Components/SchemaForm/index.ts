export * from './SchemaForm';
export * from './types';
import SchemaFormDescription from './Description';
import { getLabel as getSchemaFormLabel } from './utils';
export { SchemaFormDescription, getSchemaFormLabel };
export { default as SchemaFormField } from './Field';
