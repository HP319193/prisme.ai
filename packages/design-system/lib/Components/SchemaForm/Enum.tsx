import { useMemo } from 'react';
import { useSchemaForm } from './context';
import { FieldProps } from './types';

export const Enum = (props: FieldProps) => {
  const {
    components: { FieldSelect },
  } = useSchemaForm();
  const options = useMemo(
    () =>
      (props.schema.enum && Array.isArray(props.schema.enum)
        ? props.schema.enum
        : []
      ).map((value, k) => ({
        label:
          (props.schema.enumNames && props.schema.enumNames[k]) || `${value}`,
        value,
      })),
    [props.schema]
  );
  if (!FieldSelect) return null;
  return <FieldSelect {...props} options={options} />;
};

export default Enum;
