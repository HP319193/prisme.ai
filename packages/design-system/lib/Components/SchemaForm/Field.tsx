import { useEffect, useMemo, useRef, useState } from 'react';
import { Form, useField } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import { SchemaFormContext, useSchemaForm } from './context';
import Enum from './Enum';
import OneOf from './OneOf';
import { FieldProps } from './types';
import { getFieldOptions } from './utils';
import { OnChange } from 'react-final-form-listeners';

export interface FieldComponentProps extends FieldProps {
  components: SchemaFormContext['components'];
}

export const Field = ({ components, ...props }: FieldComponentProps) => {
  const field = useField(props.name, getFieldOptions(props.schema));

  // Update value with default value
  const [_default, setDefault] = useState();
  useEffect(() => {
    if (!_default && props.schema.default) {
      setDefault(props.schema.default);
    }
  }, [_default, props.schema.default]);
  useEffect(() => {
    if (!_default || field.input.value !== undefined) {
      return;
    }
    field.input.onChange(_default);
  }, [_default]);

  const Component = useMemo(() => {
    if (props.schema.oneOf) {
      return OneOf;
    }
    const { 'ui:widget': UiWidget } = props.schema;
    if (typeof UiWidget === 'function') {
      return UiWidget;
    }

    if (UiWidget && components.UiWidgets && components.UiWidgets[UiWidget]) {
      return components.UiWidgets[UiWidget];
    }

    // This widget can be used with any type
    if (UiWidget === 'select') {
      return components.FieldSelect;
    }

    if (props.schema.enum) {
      return Enum;
    }

    switch (props.schema.type) {
      case 'localized:string':
      case 'localized:number':
        return components.FieldLocalizedText;
      case 'string':
      case 'number':
        return components.FieldText;
      case 'boolean':
        return components.FieldBoolean;
      case 'localized:boolean':
        return components.FieldLocalizedBoolean;
      case 'object':
        return components.FieldObject;
      case 'array':
        if (UiWidget === 'tags') {
          return components.FieldTags;
        }
        if (UiWidget === 'upload') {
          return components.FieldArrayUpload;
        }
        return components.FieldArray;
      default:
        return components.FieldAny;
    }
  }, [props.schema]);

  if (props.schema.hidden || !Component) return null;

  return <Component {...props} />;
};

const LinkedField = (props: FieldProps) => {
  const { components } = useSchemaForm();
  const { Field: FieldComponent = Field } = components;
  return <FieldComponent {...props} components={components} />;
};

export const SelfField = (
  props: Pick<FieldProps, 'schema' | 'label'> & {
    value: any;
    onChange: (v: any) => void;
  }
) => {
  const values = useRef({ values: props.value });
  return (
    <Form
      onSubmit={props.onChange}
      initialValues={values.current}
      mutators={{ ...arrayMutators }}
    >
      {() => (
        <>
          <OnChange name="values">
            {(value, previous) => {
              if (previous === value) return;
              props.onChange(value);
            }}
          </OnChange>
          <LinkedField {...props} name="values" />
        </>
      )}
    </Form>
  );
};

export default LinkedField;
