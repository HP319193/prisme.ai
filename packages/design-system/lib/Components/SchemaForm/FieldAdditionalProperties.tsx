import { DeleteOutlined } from '@ant-design/icons';
import { Tooltip } from 'antd';
import { useCallback, useEffect, useRef, useState } from 'react';
import { useField, FieldRenderProps } from 'react-final-form';
import Button from '../Button';
import Input from '../Input';
import TextArea from '../TextArea';
import { useSchemaForm } from './context';
import Field from './Field';
import { FieldProps, Schema } from './types';
import { getDefaultValue } from './utils';

interface AdditionalPropertiesProps extends FieldProps {
  field: FieldRenderProps<any, HTMLElement, any>;
}

const FreeAdditionalProperties = ({
  schema,
  field,
}: AdditionalPropertiesProps) => {
  const [value, setValue] = useState('');
  useEffect(() => {
    try {
      const json = JSON.parse(value);
      const keys = Object.keys(schema.properties || {});
      const cleanedValue = Object.keys(field.input.value || {}).reduce(
        (prev, key) =>
          keys.includes(key)
            ? {
                ...prev,
                [key]: field.input.value[key],
              }
            : prev,
        {}
      );
      field.input.onChange({
        ...cleanedValue,
        ...json,
      });
    } catch {
      //
    }
  }, [value]);

  return (
    <TextArea
      onChange={({ target: { value } }) => setValue(value)}
      defaultValue={
        typeof value === 'string' ? value : JSON.stringify(value, null, '  ')
      }
    />
  );
};

const ManagedAdditionalProperties = ({
  schema,
  field,
  name,
}: AdditionalPropertiesProps) => {
  const [value, setValue] = useState<Record<string, any>>({});
  const { locales = {} } = useSchemaForm();

  const lock = useRef(false);

  useEffect(() => {
    if (lock.current) {
      lock.current = false;
      return;
    }
    const keys = Object.keys(schema.properties || {});
    const cleanedValue = Object.keys(field.input.value || {}).reduce(
      (prev, key) =>
        keys.includes(key)
          ? {
              ...prev,
              [key]: field.input.value[key],
            }
          : prev,
      {}
    );

    lock.current = true;
    field.input.onChange({
      ...cleanedValue,
      ...value,
    });
  }, [value]);

  useEffect(() => {
    if (lock.current) {
      lock.current = false;
      return;
    }
    setValue(() => {
      const keys = Object.keys(schema.properties || {});
      const cleanedValue = Object.keys(field.input.value || {}).reduce(
        (prev, key) =>
          keys.includes(key)
            ? prev
            : {
                ...prev,
                [key]: field.input.value[key],
              },
        {}
      );
      return cleanedValue;
    });
  }, [field.input.value]);

  const addKey = useCallback(() => {
    setValue((value) =>
      value.hasOwnProperty('')
        ? value
        : {
            ...value,
            '': getDefaultValue((schema.additionalProperties as Schema).type),
          }
    );
  }, []);

  const updateKey = useCallback(
    (prevKey: string) => (newKey: string) => {
      setValue((value) =>
        Object.keys(value).reduce(
          (prev, key) => ({
            ...prev,
            [key === prevKey ? newKey : key]: value[key],
          }),
          {}
        )
      );
    },
    []
  );

  const removeKey = useCallback(
    (prevKey: string) => () => {
      setValue((value) =>
        Object.keys(value).reduce(
          (prev, key) =>
            key === prevKey
              ? prev
              : {
                  ...prev,
                  [key]: value[key],
                },
          {}
        )
      );
    },
    []
  );

  return (
    <div>
      {Object.keys(value).map((key, index) => (
        <div key={index} className="flex flex-1 flex-row items-center">
          <Input
            containerClassName="flex flex-1"
            value={key}
            onChange={({ target: { value } }) => updateKey(key)(value)}
            label={locales.propertyKey || 'Key'}
          />
          {' : '}
          {key && (
            <Field
              schema={schema.additionalProperties as Schema}
              label={
                locales.propertyValue === undefined
                  ? 'Value'
                  : locales.propertyValue
              }
              name={`${name}.${key}`}
            />
          )}

          <Button onClick={removeKey(key)}>
            <Tooltip
              title={locales.removeProperty || 'Remove'}
              placement="left"
            >
              <button>
                <DeleteOutlined />
              </button>
            </Tooltip>
          </Button>
        </div>
      ))}
      <Button onClick={addKey}>
        {locales.addProperty || 'Add a property'}
      </Button>
    </div>
  );
};

export const FieldAdditionalProperties = ({ schema, name }: FieldProps) => {
  const field = useField(name);

  if (!schema || !schema.additionalProperties) return null;

  if (schema.additionalProperties === true) {
    return (
      <FreeAdditionalProperties field={field} schema={schema} name={name} />
    );
  }

  return (
    <ManagedAdditionalProperties schema={schema} field={field} name={name} />
  );
};

export default FieldAdditionalProperties;
