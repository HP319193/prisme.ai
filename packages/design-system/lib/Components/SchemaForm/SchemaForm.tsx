import {
  Children,
  cloneElement,
  MutableRefObject,
  ReactElement,
  useCallback,
  useMemo,
  useRef,
} from 'react';
import { Form } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import Field from './Field';
import { Schema } from './types';
import Button from '../Button';
import { root } from './utils';
import { context, FieldContainer, SchemaFormContext } from './context';
import FieldAny from './FieldAny';
import FieldBoolean from './FieldBoolean';
import FieldDate from './FieldDate';
import FieldLocalizedBoolean from './FieldLocalizedBoolean';
import FieldLocalizedText from './FieldLocalizedText';
import FieldObject from './FieldObject';
import FieldArray from './FieldArray';
import FieldTags from './FieldTags';
import FieldSelect from './FieldSelect';
import FieldRadio from './FieldRadio';
import FieldText from './FieldText';
import { OnChange } from 'react-final-form-listeners';
import { FormApi } from 'final-form';

export interface SchemaFormProps {
  schema: Schema;
  onSubmit?: (
    values: any
  ) => void | Record<string, any> | Promise<Record<string, any>>;
  buttons?: ReactElement[];
  onChange?: (values: any) => void;
  initialValues?: any;
  locales?: SchemaFormContext['locales'];
  components?: Partial<SchemaFormContext['components']>;
  utils?: Partial<SchemaFormContext['utils']>;
  formRef?: MutableRefObject<FormApi<any, any>>;
  initialFieldObjectVisibility?: boolean;
}

const DefaultLocales = {};

export const SchemaForm = ({
  schema,
  onSubmit,
  buttons,
  onChange,
  initialValues,
  locales = DefaultLocales,
  components,
  utils,
  formRef,
  initialFieldObjectVisibility = true,
}: SchemaFormProps) => {
  if (!schema) return null;
  const values = useRef({ values: initialValues });

  const onSubmitHandle = useCallback(
    async (values: any) => {
      if (!onSubmit) return;
      const errors = await onSubmit(values.values);
      if (!errors) return;
      return { values: errors };
    },
    [onSubmit]
  );
  const componentsWithDefault = useMemo(
    () => ({
      FieldContainer,
      FieldArray,
      FieldAny,
      FieldBoolean,
      FieldDate,
      FieldLocalizedBoolean,
      FieldLocalizedText,
      FieldObject: FieldObject(initialFieldObjectVisibility),
      FieldTags,
      FieldSelect,
      FieldRadio,
      FieldText,
      UiWidgets: {},
      ...components,
    }),
    [components, initialFieldObjectVisibility]
  );

  const utilsWithDefault = useMemo(
    () => ({
      extractSelectOptions: () => [],
      extractAutocompleteOptions: () => [],
      uploadFile: async (file: string) => file,
      ...utils,
    }),
    [utils]
  );

  return (
    <context.Provider
      value={{
        locales,
        components: componentsWithDefault,
        utils: utilsWithDefault,
      }}
    >
      <Form
        onSubmit={onSubmitHandle}
        initialValues={values.current}
        mutators={{ ...arrayMutators }}
      >
        {({ handleSubmit, hasValidationErrors, form }) => {
          formRef && (formRef.current = form);
          return (
            <form
              onSubmit={handleSubmit}
              className={`pr-form ${
                hasValidationErrors ? 'pr-form--has-validation-errors' : ''
              }`}
            >
              {onChange && (
                <OnChange name="values">
                  {(value, previous) => {
                    if (previous === value) return;
                    onChange(value);
                  }}
                </OnChange>
              )}
              <Field schema={schema} name={root} />
              {buttons ? (
                Children.map(buttons, (button, key) =>
                  cloneElement(button, {
                    key: button.key || key,
                  })
                )
              ) : (
                <Button
                  type="submit"
                  className="pr-form-submit"
                  disabled={hasValidationErrors}
                >
                  {locales.submit || 'Submit'}
                </Button>
              )}
            </form>
          );
        }}
      </Form>
    </context.Provider>
  );
};

export default SchemaForm;
