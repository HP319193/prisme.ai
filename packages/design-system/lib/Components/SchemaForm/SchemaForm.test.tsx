import { Form } from 'react-final-form';
import renderer, { act } from 'react-test-renderer';
import SchemaForm from './SchemaForm';
import { Schema } from './types';

it('should render empty form', () => {
  const schema = {};
  const root = renderer.create(<SchemaForm schema={schema} />);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should submit form', () => {
  const schema = {};
  const onSubmit = jest.fn();
  const root = renderer.create(
    <SchemaForm schema={schema} onSubmit={onSubmit} />
  );
  const expected = {};
  act(() => {
    root.root.findByType(Form).props.onSubmit({ values: expected });
  });
  expect(onSubmit).toHaveBeenCalledWith(expected);
});

it('should change form', () => {
  const schema = {};
  const onChange = jest.fn();
  const root = renderer.create(
    <SchemaForm schema={schema} onChange={onChange} />
  );
  const expected = {};
  act(() => {
    root.root
      .findByProps((props: any) => !!props.onChange)
      .props.onChange(expected);
  });
  expect(onChange).toHaveBeenCalledWith(expected);
});

it('should render buttons', () => {
  const schema = {};
  const root = renderer.create(
    <SchemaForm
      schema={schema}
      buttons={[<button>foo</button>, <button>bar</button>]}
    />
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it('should render no button', () => {
  const schema = {};
  const root = renderer.create(<SchemaForm schema={schema} buttons={[]} />);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should render a single input', () => {
  const schema: Schema = {
    type: 'string',
  };
  const root = renderer.create(<SchemaForm schema={schema} buttons={[]} />);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should render single all types', () => {
  const schema: Schema = {
    type: 'object',
    properties: {
      string: {
        type: 'string',
      },
      number: {
        type: 'number',
      },
      boolean: {
        type: 'boolean',
      },
      array: {
        type: 'array',
        title: 'Localized',
        items: {
          type: 'object',
          properties: {
            string: {
              type: 'localized:string',
            },
            number: {
              type: 'localized:number',
            },
            boolean: {
              type: 'localized:boolean',
            },
          },
        },
      },
      object: {
        type: 'object',
        additionalProperties: true,
      },
      objectWithAdditionalProperties: {
        type: 'object',
        additionalProperties: {
          type: 'object',
          properties: {
            foo: {
              type: 'string',
            },
          },
        },
      },
    },
  };
  const root = renderer.create(<SchemaForm schema={schema} buttons={[]} />);
  expect(root.toJSON()).toMatchSnapshot();
});

it('should set locales', () => {
  const schema: Schema = {
    type: 'string',
  };
  const root = renderer.create(
    <SchemaForm
      schema={schema}
      locales={{
        submit: 'SUBMIT',
      }}
    />
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it('should set custom components', () => {
  const schema: Schema = {
    type: 'string',
  };
  const FieldText = jest.fn(() => <div>Custom FieldText</div>);
  const root = renderer.create(
    <SchemaForm
      schema={schema}
      components={{
        FieldText,
      }}
    />
  );
  expect(root.toJSON()).toMatchSnapshot();
  expect(FieldText).toHaveBeenCalledWith(
    expect.objectContaining({ schema, name: 'values' }),
    {}
  );
});

it('should use ui:widgets', () => {
  const schema: Schema = {
    type: 'object',
    properties: {
      array: {
        type: 'array',
        'ui:options': {
          array: 'row',
        },
      },
      upload: {
        type: 'string',
        'ui:widget': 'upload',
        'ui:options': {
          upload: { accept: 'image/jpeg' },
        },
      },
      textarea: {
        type: 'string',
        'ui:widget': 'textarea',
        'ui:options': {
          textarea: { rows: 6 },
        },
      },
      select: {
        type: 'string',
        'ui:widget': 'select',
        'ui:options': {
          select: {
            options: [
              { label: 'Foo', value: 'foo' },
              { label: 'Bar', value: 'bar' },
            ],
          },
        },
      },
      radio: {
        type: 'string',
        'ui:widget': 'radio',
        'ui:options': {
          select: {
            options: [
              { label: 'Foo', value: 'foo' },
              { label: 'Bar', value: 'bar' },
            ],
          },
        },
      },
      date: {
        type: 'string',
        'ui:widget': 'date',
        'ui:options': {
          date: {
            showToday: false,
          },
        },
      },
    },
  };
  const root = renderer.create(<SchemaForm schema={schema} />);
  expect(root.toJSON()).toMatchSnapshot();
});
