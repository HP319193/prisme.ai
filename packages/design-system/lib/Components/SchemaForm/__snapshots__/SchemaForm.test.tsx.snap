// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a single input 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div
    className="pr-form-field pr-form-text"
  />
</form>
`;

exports[`should render buttons 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div
    className="pr-form-field pr-form-any"
  />
  <button>
    foo
  </button>
  <button>
    bar
  </button>
</form>
`;

exports[`should render empty form 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div
    className="pr-form-field pr-form-any"
  />
  Submit
</form>
`;

exports[`should render no button 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div
    className="pr-form-field pr-form-any"
  />
</form>
`;

exports[`should render single all types 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div
    className="pr-form-field pr-form-object"
  >
    
    <div
      className="transition-[height] overflow-hidden will-change-contents "
    >
      <div>
        <div
          className="pr-form-object__properties"
        >
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-text"
            >
              <label
                className="pr-form-text__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "string",
                  }
                }
                htmlFor="values.string"
              />
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-text"
            >
              <label
                className="pr-form-text__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "number",
                  }
                }
                htmlFor="values.number"
              />
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-boolean"
            >
              <label
                className="pr-form-boolean__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "boolean",
                  }
                }
                htmlFor="values.boolean"
              />
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-array"
            >
              <label
                className="pr-form-array__label pr-form-label"
                htmlFor="values.array"
              >
                Localized
              </label>
              <div
                className="pr-form-array__item "
              />
              <div
                className="pr-form-array__item-add"
              />
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-object"
            >
              <button
                className="pr-form-object__label pr-form-label pr-form-object__label--visible"
                onClick={[Function]}
                type="button"
              >
                <span>
                  object
                </span>
              </button>
              <div
                className="transition-[height] overflow-hidden will-change-contents "
              >
                <div>
                  <div
                    className="pr-form-object__properties"
                  >
                    <div
                      className="pr-form-field pr-form-additional-properties pr-form-additional-properties--free"
                    >
                      <div
                        className="flex flex-1 flex-col justify-between "
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-object"
            >
              <button
                className="pr-form-object__label pr-form-label pr-form-object__label--visible"
                onClick={[Function]}
                type="button"
              >
                <span>
                  objectWithAdditionalProperties
                </span>
              </button>
              <div
                className="transition-[height] overflow-hidden will-change-contents "
              >
                <div>
                  <div
                    className="pr-form-object__properties"
                  >
                    <div
                      className="pr-form-field pr-form-additional-properties"
                    >
                      Add a property
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`should set custom components 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div>
    Custom FieldText
  </div>
  Submit
</form>
`;

exports[`should set locales 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div
    className="pr-form-field pr-form-text"
  />
  SUBMIT
</form>
`;

exports[`should use ui:widgets 1`] = `
<form
  className="pr-form "
  onSubmit={[Function]}
>
  <div
    className="pr-form-field pr-form-object"
  >
    
    <div
      className="transition-[height] overflow-hidden will-change-contents "
    >
      <div>
        <div
          className="pr-form-object__properties"
        >
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-array"
            >
              <label
                className="pr-form-array__label pr-form-label"
                htmlFor="values.array"
              >
                array
              </label>
              <div
                className="pr-form-array__item pr-form-array__item--as-row"
              />
              <div
                className="pr-form-array__item-add"
              />
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-upload"
            >
              <label
                className="pr-form-upload__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "upload",
                  }
                }
                htmlFor="values.upload"
              />
              <div
                className="pr-form-upload__input pr-form-input"
              >
                <div
                  className="pr-form-upload__placeholder"
                >
                  <div
                    className="pr-form-upload__preview"
                  />
                  Choose file
                </div>
                <input
                  accept="image/jpeg"
                  multiple={false}
                  onChange={[Function]}
                  type="file"
                />
                
              </div>
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-text pr-form-text--textarea"
            >
              <label
                className="pr-form-text__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "textarea",
                  }
                }
                htmlFor="values.textarea"
              />
              <div
                className="pr-form-text__input pr-form-input"
              />
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-select"
            >
              <label
                className="pr-form-select__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "select",
                  }
                }
                htmlFor="values.select"
              />
              Foo
              Bar
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-text"
            >
              <label
                className="pr-form-text__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "radio",
                  }
                }
                htmlFor="values.radio"
              />
            </div>
          </div>
          <div
            className="pr-form-object__property"
          >
            <div
              className="pr-form-field pr-form-date"
            >
              <label
                className="pr-form-date__label pr-form-label"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "date",
                  }
                }
                htmlFor="values.date"
              />
              <div
                className="pr-form-date__input pr-form-input"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  Submit
</form>
`;
